hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/PostponedSnapshotTest.java

```diff
@@ -0,0 +1,114 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.core;
+
+import com.hazelcast.jet.JetInstance;
+import com.hazelcast.jet.JetTestInstanceFactory;
+import com.hazelcast.jet.Job;
+import com.hazelcast.jet.config.JobConfig;
+import com.hazelcast.jet.impl.SnapshotRepository;
+import com.hazelcast.jet.impl.execution.SnapshotRecord;
+import com.hazelcast.jet.impl.execution.SnapshotRecord.SnapshotStatus;
+import com.hazelcast.jet.stream.IStreamMap;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.concurrent.atomic.AtomicIntegerArray;
+
+import static com.hazelcast.jet.core.Edge.between;
+import static com.hazelcast.jet.core.Edge.from;
+import static com.hazelcast.jet.core.JetTestSupport.assertTrueEventually;
+import static com.hazelcast.jet.core.processor.DiagnosticProcessors.writeLogger;
+import static com.hazelcast.jet.impl.execution.SnapshotRecord.SnapshotStatus.ONGOING;
+import static com.hazelcast.jet.impl.execution.SnapshotRecord.SnapshotStatus.SUCCESSFUL;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class PostponedSnapshotTest {
+
+    private static volatile AtomicIntegerArray latches;
+
+    private JetInstance instance;
+    private JetTestInstanceFactory factory;
+
+    @Before
+    public void setup() {
+        factory = new JetTestInstanceFactory();
+        instance = factory.newMember();
+        latches = new AtomicIntegerArray(2);
+    }
+
+    @After
+    public void tearDown() {
+        factory.shutdownAll();
+    }
+
+    @Test
+    public void when_jobHasHigherPriorityEdge_then_noSnapshotUntilEdgeDone() throws InterruptedException {
+        DAG dag = new DAG();
+        Vertex highPrioritySource = dag.newVertex("highPrioritySource", () -> new SourceP(0));
+        Vertex lowPrioritySource = dag.newVertex("lowPrioritySource", () -> new SourceP(1));
+        Vertex sink = dag.newVertex("sink", writeLogger());
+
+        dag.edge(between(highPrioritySource, sink).priority(-1))
+           .edge(from(lowPrioritySource).to(sink, 1));
+
+        JobConfig config = new JobConfig();
+        config.setSnapshotIntervalMillis(100);
+        Job job = instance.newJob(dag, config);
+
+        IStreamMap<Object, Object> snapshotsMap = instance.getMap(SnapshotRepository.snapshotsMapName(job.getJobId()));
+
+        // check, that snapshot starts, but stays in ONGOING state
+        assertTrueEventually(() -> assertTrue(existsSnapshot(snapshotsMap, ONGOING)), 5);
+        Thread.sleep(1000);
+        assertFalse(existsSnapshot(snapshotsMap, SUCCESSFUL));
+
+        latches.set(0, 1);
```

Comment:
I'd prefer to use two separate latches here, and then have just a single CountdownLatch passed into SourceP.
Looks good otherwise.

Reply:
You can't pass `CountdownLatch` or `AtomicInteger` through serialization. It has to be shared static field.

Reply:
ok, I see your point.